packages/automerge-repo/src/DocHandle.ts

```diff
@@ -78,156 +63,92 @@ export class DocHandle<T> //
-                {
-                  delay: this.#timeoutDelay,
-                  target: UNAVAILABLE,
-                },
-              ],
-            },
-            ready: {
-              on: {
-                // UPDATE is called by the Repo when we receive changes from the network
-                UPDATE: { actions: "onUpdate", target: READY },
-                DELETE: { actions: "onDelete", target: DELETED },
-              },
-            },
-            deleted: {
-              type: "final",
-            },
-            unavailable: {
-              on: {
-                UPDATE: { actions: "onUpdate" },
-                // REQUEST_COMPLETE is called from `onUpdate` when the doc has been fully loaded from the network
-                REQUEST_COMPLETE: { target: READY },
-                DELETE: { actions: "onDelete", target: DELETED },
-              },
-            },
-          },
+    const delay = this.#timeoutDelay
+    const machine = setup({
+      types: {
+        context: {} as DocHandleContext<T>,
+        events: {} as DocHandleEvent<T>,
+      },
+      actions: {
+        /** Update the doc using the given callback and put the modified doc in context */
+        onUpdate: assign(({ context, event }) => {
+          const oldDoc = context.doc
+
+          assertEvent(event, UPDATE)
+          const { callback } = event.payload
+          const newDoc = callback(oldDoc)
+
+          return { doc: newDoc }
+        }),
+        onDelete: assign(() => {
+          this.emit("delete", { handle: this })
+          return { doc: undefined }
+        }),
+        onUnavailable: () => {
+          this.emit("unavailable", { handle: this })
         },
+      },
+    }).createMachine({
+      /** @xstate-layout N4IgpgJg5mDOIC5QAoC2BDAxgCwJYDswBKAYgFUAFAEQEEAVAUQG0AGAXUVAAcB7WXAC64e+TiAAeiAOwAOAKwA6ACxSAzKqks1ATjlTdAGhABPRAFolAJksKN2y1KtKAbFLla5AX09G0WPISkVAwAMgyMrBxIILz8QiJikggAjCzOijKqLEqqybJyLizaRqYIFpbJtro5Uo7J2o5S3r4YOATECrgQADZgJADCAEoM9MzsYrGCwqLRSeoyCtra8pa5adquySXmDjY5ac7JljLJeepKzSB+bYGdPX0AYgCSAHJUkRN8UwmziM7HCgqyVcUnqcmScmcMm2ZV2yiyzkOx1OalUFx8V1aAQ63R46AgBCgJGGAEUyAwAMp0D7RSbxGagJKHFgKOSWJTJGRSCosCpKaEmRCqbQKU5yXINeTaer6LwY67YogKXH4wkkKgAeX6AH1hjQqABNGncL70xKIJQ5RY5BHOJag6wwpRyEWImQVeT1aWrVSXBXtJUqgn4Ik0ADqNCedG1L3CYY1gwA0saYqbpuaEG4pKLksKpFDgcsCjDhTnxTKpTLdH6sQGFOgAO7oKYhl5gAQNngAJwA1iRY3R40ndSNDSm6enfpm5BkWAVkvy7bpuTCKq7ndZnfVeSwuTX-HWu2AAI4AVzgQhD6q12rILxoADVIyEaAAhMLjtM-RmIE4LVSQi4nLLDIGzOCWwLKA0cgyLBoFWNy+43B0R5nheaqajqepjuMtJfgyEh-FoixqMCoKqOyhzgYKCDOq6UIeuCSxHOoSGKgop74OgABuzbdOgABGvTXlho5GrhJpxJOP4pLulT6KoMhpJY2hzsWNF0QobqMV6LG+pc+A8BAcBiP6gSfFJ36EQgKksksKxrHamwwmY7gLKB85QjBzoAWxdZdL0FnfARST8ooLC7qoTnWBU4pyC5ViVMKBQaHUDQuM4fm3EGhJBWaU7-CysEAUp3LpEpWw0WYRw2LmqzgqciIsCxWUdI2zaXlAbYdt2PZ5dJ1n5jY2iJY1ikOIcMJHCyUWHC62hRZkUVNPKta3Kh56wJ1-VWUyzhFc64JWJCtQNBBzhQW4cHwbsrVKpxPF8YJgV4ZZIWIKkiKiiNSkqZYWjzCWaQ5hFh0AcCuR3QoR74qUknBRmzholpv3OkpRQNNRpTzaKTWKbIWR5FDxm9AIkA7e9skUYCWayLILBZGoLkUSKbIyIdpxHPoyTeN4QA */
```

Comment:
This blob is automatically updated by the XState extension for VS Code extension.